sdk/spring/spring-messaging-azure-eventhubs/src/test/java/com/azure/spring/eventhubs/core/EventHubsTemplateTest.java

```diff
@@ -0,0 +1,192 @@
+     */
+    @Test
+    void testSendAsyncForMessagesTwoBatchAndSendCompletelyWithException() {
+        EventDataBatch eventDataBatch = mock(EventDataBatch.class);
+
+        when(this.mockProducerClient.createBatch(any(CreateBatchOptions.class)))
+            .thenReturn(Mono.just(eventDataBatch));
+        when(eventDataBatch.tryAdd(any(EventData.class))).thenReturn(true, false, true).
+                                                         thenThrow(new AmqpException(false,
+                                                             AmqpErrorCondition.LINK_PAYLOAD_SIZE_EXCEEDED,
+                                                             String.format(Locale.US, "Size of the payload exceeded "
+                                                                     + "maximum message size: %s kb",
+                                                                 1024 * 1024 / 1024),
+                                                             null)).thenReturn(true, true);
+        when(eventDataBatch.getCount()).thenReturn(5);
+        List<String> messagesList = new ArrayList<>();
+        for (int i = 0; i < 5; i++) {
+            messagesList.add("abcde");
+        }
+        List<Message<String>> messages =
+            messagesList.stream().map((Function<String, GenericMessage<String>>) GenericMessage::new).collect(Collectors.toList());
+
+        this.eventHubsTemplate.sendAsync(this.destination, messages, null).doOnError(ex -> {
+            System.out.println("do on Error");
+            ex.printStackTrace();
+        }).doOnSuccess(t -> {
+            System.out.println("do on success:" + t);
+        }).block(Duration.ofSeconds(10));
+        verify(this.mockProducerClient, times(2)).send(any(EventDataBatch.class));
+    }
+
+    /**
+     * test the case that the second event is too large for one new batch
+     */
+    @Test
+    void testSendAsyncForMessagesWithTheSecondEventTooLargeForOneNewBatch() {
+        EventDataBatch eventDataBatch = mock(EventDataBatch.class);
+
+        when(this.mockProducerClient.createBatch(any(CreateBatchOptions.class)))
+            .thenReturn(Mono.just(eventDataBatch));
+        when(eventDataBatch.tryAdd(any(EventData.class))).thenReturn(true, false, false).thenReturn(true, true, true);
+        when(eventDataBatch.getCount()).thenReturn(5);
+        List<String> messagesList = new ArrayList<>();
+        for (int i = 0; i < 5; i++) {
+            messagesList.add("abcde");
+        }
+        List<Message<String>> messages =
+            messagesList.stream().map((Function<String, GenericMessage<String>>) GenericMessage::new).collect(Collectors.toList());
+
+        this.eventHubsTemplate.sendAsync(this.destination, messages, null).doOnError(ex -> {
+            System.out.println("do on Error" + ex.getMessage());
+            ex.printStackTrace();
+        }).doOnSuccess(t -> {
+            System.out.println("do on success:" + t);
+        }).block(Duration.ofSeconds(10));
```

Comment:
Can we use https://projectreactor.io/docs/test/release/api/reactor/test/StepVerifier.html here?

Reply:
ok, update adding reactor-test dependency and replace 'block' method with 'verifyComlete' from StepVerifier.